src/Connection.php

```diff
@@ -1094,23 +1098,46 @@ public function getTransactionNestingLevel(): int
     }
 
     /**
-     * Returns the ID of the last inserted row, or the last value from a sequence object,
-     * depending on the underlying driver.
+     * Returns the ID of the last inserted row.
      *
-     * Note: This method may not return a meaningful or consistent result across different drivers,
-     * because the underlying database may not even support the notion of AUTO_INCREMENT/IDENTITY
-     * columns or sequences.
+     * If the underlying driver does not support identity columns, an exception is thrown.
      *
-     * @param string|null $name Name of the sequence object from which the ID should be returned.
+     * @return int|string The last insert ID, as an integer or a numeric string.
      *
-     * @return string A string representation of the last inserted ID.
+     * @throws Exception
+     */
+    public function lastInsertId()
+    {
+        try {
+            return $this->getWrappedConnection()->lastInsertId();
+        } catch (DriverException\IdentityColumnsNotSupported $e) {
+            throw new IdentityColumnsNotSupported($e->getMessage(), $e);
+        } catch (DriverException\NoIdentityValue $e) {
+            throw new NoIdentityValue($e->getMessage(), $e);
+        } catch (DriverException $e) {
+            throw $this->convertException($e);
+        }
+    }
+
+    /**
+     * Returns the current sequence value for the given sequence name.
+     *
+     * If the underlying driver does not support sequences, an exception is thrown.
+     *
+     * @param string $name The sequence name.
+     *
+     * @return int|string The sequence number, as an integer or a numeric string.
      *
      * @throws Exception
      */
-    public function lastInsertId(?string $name = null): string
+    public function getSequenceNumber(string $name)
     {
         try {
-            return $this->getWrappedConnection()->lastInsertId($name);
+            return $this->getWrappedConnection()->getSequenceNumber($name);
+        } catch (DriverException\SequencesNotSupported $e) {
+            throw new SequencesNotSupported($e->getMessage(), $e);
+        } catch (DriverException\SequenceDoesNotExist $e) {
```

Comment:
Same as above, the specialized wrapper-level exception classes are redundant IMO.

Reply:
Do we need driver-level exception classes then? We could just throw a generic driver exception, because now they don't need to be caught.

Reply:
> Do we need driver-level exception classes then?

1. Testability. We can expect a certain exception class in a certain case.
2. Ease of construction. The named constructor can accept exception-specific parameters and encapsulate the creation of the exception message.

Note that all driver-level exceptions are marked as `@internal`, so if you try to catch a specific class in your code, some code analyzers will tell you that you're doing it wrong.

Reply:
> 1. Testability. We can expect a certain exception class in a certain case.

<s>As far as testability is concerned, this would only be useful if we specifically tested each driver in isolation; as it stands, tests target the wrapper, so we have no way to catch the driver-level exception.</s> Actually we do: we can test the previous exceptions and see if any of them is the driver-level exception. Is that acceptable in tests?

> 2. Ease of construction. The named constructor can accept exception-specific parameters and encapsulate the creation of the exception message.

If this is the only justification, we can also create factory methods on a generic driver-level class for these use cases, if we want to avoid creating exceptions that will never be caught anywhere.

> Note that all driver-level exceptions are marked as @internal, so if you try to catch a specific class in your code, some code analyzers will tell you that you're doing it wrong.

You should have no way to catch driver-level exceptions anyway, as they'll will have been converted to wrapper-level exceptions before they can reach userland code. Unless I missed something?

Reply:
> [...] we can test the previous exceptions and see if any of them is the driver-level exception. Is that acceptable in tests?

As long as there no dedicated wrapper-level exception, I think yes. Some tests work with the drivers directly, it may be even better if we want to not run the code that is not part of the behavior being tested.

> If this is the only justification, we can also create factory methods on a generic driver-level class for these use cases, if we want to avoid creating exceptions that will never be caught anywhere.

How is the exception produced by a factory method meant to be identified? Also, the single responsibility principle that the generic driver-level class will violate.

> You should have no way to catch driver-level exceptions anyway, as they'll will have been converted to wrapper-level exceptions before they can reach userland code. Unless I missed something?

Right, but if not marked internal explicitly they are automatically part of the public API of the library, but I want to be flexible in changing how the exceptions are constructed w/o the consumers of the API being impacted by the changes.

Reply:
> How is the exception produced by a factory method meant to be identified?

That's the whole point: if we don't catch the specific driver exception (`SequencesNotSupported`, for example), the exception does not need to be identified;.

> Also, the single responsibility principle that the generic driver-level class will violate.

I don't think factory methods are a violation of the SRP, they all create a driver exception with a specific error message.

But, anyway, I think keeping internal exceptions for identity/sequences is a good idea.

> Right, but if not marked internal explicitly they are automatically part of the public API of the library, but I want to be flexible in changing how the exceptions are constructed w/o the consumers of the API being impacted by the changes.

Sure, I will remove the public exceptions:

- remove the wrapper-level exceptions
- make driver-level exceptions `@internal`
- change the tests so that they check the previous exceptions looking for the internal driver exception

Reply:
> if we don't catch the specific driver exception (`SequencesNotSupported`, for example), the exception does not need to be identified

It needs to be identified for testing. I want to make sure that the driver exception that occurs during a test is exactly the `SequenceDoesNotExist` exception, not whatever connection/syntax error.

As for `SequencesNotSupported`, it's a totally different story. The fact that the user finds out that they cannot use a certain API only at runtime is a design smell. The `AbstractPlatform` class contains ~200 public methods which makes it quite a god object, and this is a side effect of that.

Reply:
Removed the public exceptions in b6f1c06f6552e69917dc9adb8ab7f3522b65394e, please tell me if you're happy with the `expectPreviousException()` thing.